### PR TITLE
Shape ops without copies

### LINKED ISSUE #1860: `fold` does not fail if requested input dim is not found
Instead it just returns the object unchanged (copied). Can likely become a source of bugs. Raise an exception instead.

### LINKED ISSUE #1863: `Variable::setDims()` may have wrong stride setup if input is not contiguous
This was probably not an issue previously because variables where always contiguous, but with the sharing rewrite this is not the case any more.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Fixes #1860.
- `broadcast` and `fold` now return views
- `flatten` returns views if possible.
- remove `reshape` since it is mostly covered by `fold` and `flatten` and would require more complicated handling/validation of dims and strides.
- `ones` does *not* return a broadcast scalar 1.
- Move comparison and transform tests out of `variable_test.cpp`.
- Add tests for `copy(Variable)`.
- Remove `Variable::setDims()` to fix #1863.

### CODE REVIEW

File: variable/include/scipp/variable/variable.h [hunk 1]
```diff
@@ -139,6 +141,9 @@ class SCIPP_VARIABLE_EXPORT Variable {
 
   [[nodiscard]] Variable as_const() const;
 
+  auto &unchecked_dims() { return m_dims; }
+  auto &unchecked_strides() { return m_strides; }
```
Comment (SimonHeybrock) [inline variable/include/scipp/variable/variable.h:145]: I don't like this. Does anyone have a nicer idea, apart from making all functions which need to access this members? @jl-wynen

**Comment by SimonHeybrock** [comment]
Marked as draft because while it looks coorect in a notebook, `MultiIndex` cannot handle stride setup with overlaps right now, this should be fixed with a refactor to remove `m_dataDims` that @jl-wynen is working on.

File: variable/include/scipp/variable/variable.h (hunk 1 above)
Comment (jl-wynen) [reply variable/include/scipp/variable/variable.h:145]: Are they really needed? The check is not particularly expensive.

Comment (SimonHeybrock) [reply variable/include/scipp/variable/variable.h:145]: Which check? The point is that we have no way of setting strides or dims right now, without (in this case) unintended side effects (such as resizing the buffer).

Comment (SimonHeybrock) [reply variable/include/scipp/variable/variable.h:145]: Basically, there is a different check for every shape-related operation (such as broadcast, fold, flatten, reshape).

Comment (jl-wynen) [reply variable/include/scipp/variable/variable.h:145]: Ah, I though you wanted to circumvent the `!*this` check in `dims()`.

The best I can think of right now is a setter with a scary sounding name (to discourage usage). Might be a bit nicer than the getters. But not by much...

**Comment by SimonHeybrock** [comment]
@jl-wynen @nvaytet After merging the strides PR this is ready for review now, please have a look. Cheers!

**Review by jl-wynen — APPROVED** [review]
Looks good

**Pull request merged**